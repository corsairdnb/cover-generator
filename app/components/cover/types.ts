import { Asset } from '../../editor/Asset';

export type EditorCallbacks = {
  onUpdateImageUrl: (dataUrl: string) => void;
  onUpdateImage: (image?: HTMLImageElement) => void;
  onUpdateFile: (file?: File) => void;
};

export type Preset = {
  assets: Asset[];
};
