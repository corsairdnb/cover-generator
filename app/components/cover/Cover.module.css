.wrapper {
  display: grid;
  font-size: 1rem;
  grid-template-areas: "left right";
  grid-template-columns: 50vw 50vw;
  height: 100%;
}

.left {
  grid-area: left;
  /*background: red;*/
}

.right {
  grid-area: right;
  /*background: blue;*/
}

.cover {

}
