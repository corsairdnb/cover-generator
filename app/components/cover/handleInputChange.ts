import { SyntheticEvent } from 'react';
import { Editor } from '../../editor/Editor';
import { EditorCallbacks } from './types';

export const handleInputChange = (
  changeEvent: SyntheticEvent<HTMLInputElement>,
  editor: Editor | null,
  callbacks: EditorCallbacks
) => {
  const { onUpdateImageUrl, onUpdateImage, onUpdateFile } = callbacks;
  const fileInput = changeEvent.currentTarget;

  if (!editor || !fileInput.files?.length) {
    return;
  }

  const currentFile = fileInput.files[0];

  if (!/image.*/.exec(currentFile.type)) {
    return;
  }

  const reader = new FileReader();
  reader.readAsDataURL(currentFile);
  const currentImage = new Image();
  reader.onload = (event) => {
    if (event.target?.readyState === FileReader.DONE) {
      currentImage.src = <string>event.target?.result?.toString();
      currentImage.onload = () => {
        editor.image = currentImage;
        onUpdateImageUrl(editor.getDataUrl() || '');
        onUpdateImage(currentImage);
        onUpdateFile(currentFile);
      };
    }
  };
};
