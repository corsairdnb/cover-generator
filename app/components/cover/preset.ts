// TODO: this props should be provided at application start as json file

import { InitialPreset } from './types';

export const initialPreset: InitialPreset = {
  assets: [
    {
      left: 100,
      right: 0,
      top: 0,
      bottom: 50,
      maxWidth: 200
    }
  ],
  labels: [
    {
      text: 'Hello world',
      left: 100,
      right: 0,
      bottom: 0,
      top: 100,
      maxWidth: 1000,
      fontSize: 100
    }
  ]
};
