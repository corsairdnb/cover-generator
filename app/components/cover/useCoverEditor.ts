import { RefObject, useEffect, useRef } from 'react';
import { Editor } from '../../editor/Editor';

export type Refs = {
  canvasRef: RefObject<HTMLCanvasElement>;
  fileInputRef: RefObject<HTMLInputElement>;
  editorRef: RefObject<Editor>;
};

export const useCoverEditor = (): Refs => {
  const canvasRef = useRef<HTMLCanvasElement | null>(null);
  const editorRef = useRef<Editor | null>(null);
  const fileInputRef = useRef<HTMLInputElement | null>(null);

  useEffect(() => {
    const canvas = canvasRef.current;
    let context: CanvasRenderingContext2D | undefined | null = null;
    if (canvas) {
      context = canvas.getContext('2d');
    }
    if (!editorRef.current && context) {
      editorRef.current = new Editor(context);
    }
  }, [canvasRef.current]);

  return { canvasRef, fileInputRef, editorRef };
};
