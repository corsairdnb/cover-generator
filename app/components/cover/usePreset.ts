import { RefObject, useEffect, useRef } from 'react';
import { ImageAsset } from '../../editor/ImageAsset';
import logo from '../../assets/logo.png';
import { Asset } from '../../editor/Asset';
import { Editor } from '../../editor/Editor';

type Preset = {
  assets?: Asset[];
};

export const usePreset = (
  editorRef: RefObject<Editor>,
  canvasRef: RefObject<HTMLCanvasElement>,
  url: string
): Preset | undefined => {
  const logoRef = useRef<string | null>(null);
  const assetsRef = useRef<Asset[]>([]);

  useEffect(() => {
    const canvas = canvasRef.current;
    const editor = editorRef.current;
    if (!editor || !canvas) return;

    if (!logoRef.current) {
      logoRef.current = logo as string;
    }

    const ctx = editor.ctx;
    const image = new Image();
    image.src = logoRef.current;

    // TODO: useMemo

    const logoAsset = new ImageAsset(ctx, image);
    const y = canvas.getBoundingClientRect().height - image.height;
    const maxWidth = 100;
    const height = Math.round((maxWidth * image.height) / image.width);
    logoAsset.attributes = new DOMRect(100, y, maxWidth, height);

    assetsRef.current = [logoAsset];
  }, [url]);

  return { assets: assetsRef.current };
};
