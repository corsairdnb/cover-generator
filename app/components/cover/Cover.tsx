import React, { FC, SyntheticEvent, useCallback, useEffect, useRef, useState } from 'react';
import { debounce } from 'throttle-debounce';
import { useCoverEditor } from './useCoverEditor';
import { handleInputChange } from './handleInputChange';
import { debounceTime } from './constants';
import styles from './Cover.module.scss';
import { usePreset } from './usePreset';

export const Cover: FC = () => {
  const [imageDataUrl, setImageDataUrl] = useState('');
  const [image, setImage] = useState<HTMLImageElement | undefined>();
  const [file, setFile] = useState<File | undefined>();
  const coverContainerRef = useRef<HTMLDivElement | null>(null);
  const { canvasRef, fileInputRef, editorRef } = useCoverEditor({ coverContainerRef });
  const preset = usePreset(editorRef, canvasRef);

  useEffect(() => {
    const editor = editorRef.current;
    if (!editor) return;
    preset
      .then(({ assets, labels }) => {
        editor.addAssets(assets);
        editor.addLabels(labels);
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.log(error);
      });
  }, []);

  const onImageChange = useCallback(
    (event: SyntheticEvent<HTMLInputElement>) =>
      handleInputChange(event, editorRef.current, {
        onUpdateImageUrl: setImageDataUrl,
        onUpdateImage: setImage,
        onUpdateFile: setFile
      }),
    [editorRef.current]
  );

  const setNewDataUrl = useCallback(() => {
    const editor = editorRef.current;
    if (!editor) return;
    setImageDataUrl(editor.getDataUrl());
  }, [editorRef.current]);

  const fontInputCallback = useCallback(
    (value: string) => {
      const editor = editorRef.current;
      if (!editor) return;
      editor.font = value;
      setNewDataUrl();
    },
    [editorRef.current]
  );

  const onFontFamilyInput = debounce(debounceTime, (value: string) => {
    fontInputCallback(value);
  });

  return (
    <div className={styles.wrapper}>
      <div className={styles.left}>
        <div className={styles.content}>
          <p>
            Date: <input type="text" placeholder="31/12" />
          </p>
          <p>
            Time: <input type="text" placeholder="19:00" />
          </p>
          <p>
            Program: <input type="text" placeholder="Xtra" />
          </p>
          <p>
            Artists: <textarea placeholder="Name..." />
          </p>
        </div>
        <div className={styles.design}>
          <p>
            Logo:{' '}
            <select>
              <option>Logo 1</option>
              <option>Logo 2</option>
              <option>Logo 3</option>
            </select>
          </p>
          <p>
            Choose File: <input type="file" id="file" ref={fileInputRef} onChange={onImageChange} />
          </p>
          {image && (
            <p>
              Image dimensions: {image.width}x{image.height}
            </p>
          )}
          {file && (
            <p>
              <a href={imageDataUrl} download={file.name}>
                Download image
              </a>
            </p>
          )}
          <p>
            Font family:{' '}
            <input
              type="text"
              placeholder="Arial"
              onInput={({ currentTarget: { value } }) => onFontFamilyInput(value)}
            />
          </p>
        </div>
      </div>
      <div className={styles.right} ref={coverContainerRef}>
        <canvas id="canvas" className={styles.cover} ref={canvasRef} />
      </div>
    </div>
  );
};
