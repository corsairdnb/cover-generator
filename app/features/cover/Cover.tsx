import React, { FC, useCallback, useEffect, useRef, useState } from 'react';
import classnames from 'classnames';
import { useCoverEditor } from './useCoverEditor';
import { usePreset } from './usePreset';
import { useDate } from './hooks/useDate';
import { useFontFamily } from './hooks/useFontFamily';
import { useTime } from './hooks/useTime';
import { useProgram } from './hooks/useProgram';
import { useArtist } from './hooks/useArtist';
import { useImage } from './hooks/useImage';
import { Label, LabelProps } from '../../editor/Label';
import styles from './Cover.module.scss';
import { useContent } from './hooks/useContent';
import { programs } from './preset';
import { Asset, AssetProps, VerticalAlignment } from '../../editor/Asset';

export const Cover: FC = () => {
  const [imageDataUrl, setImageDataUrl] = useState('');
  const [image, setImage] = useState<HTMLImageElement | undefined>();
  const coverContainerRef = useRef<HTMLDivElement | null>(null);
  const { canvasRef, fileInputRef, editorRef } = useCoverEditor({ coverContainerRef });
  const preset = usePreset(editorRef, canvasRef);
  const content = useContent();
  const { firstLine, fileName, date, time, program, artist, fontFamily, color } = content;

  const onUpdate = useCallback(() => {
    const editor = editorRef.current;
    if (!editor) return;
    const labelProps: LabelProps = {
      id: 'firstLine',
      text: firstLine,
      textAfter: '',
      left: 130,
      top: 140,
      bottom: 0,
      right: 0,
      fontSize: 76,
      maxWidth: 0,
      color: '#fff'
    };
    const logoProps: AssetProps = {
      left: 110,
      top: 0,
      bottom: 105,
      right: 0,
      maxWidth: 0,
      width: 190,
      maxHeight: 0,
      verticalAlignment: VerticalAlignment.CENTER
    };
    const logoImage = new Image();
    const programLogo = require(`../../assets/programs/${program}.png`) as { default: string };
    logoImage.src = programLogo.default;
    logoImage.onload = () => {
      editor.addAssets([new Asset(logoProps, logoImage, 'program')]);
    };
    editor.setLabels([new Label(labelProps, labelProps.id)]);
    editor.logoColor = color;
    setImageDataUrl(editor.getDataUrl());
  }, [editorRef.current, firstLine, artist, fontFamily, color, program]);

  const onFieldChange = useCallback(() => {
    const editor = editorRef.current;
    if (!editor) return;
    onUpdate();
  }, [editorRef.current, onUpdate]);

  useEffect(() => {
    if (!editorRef.current) return;
    onUpdate();
  }, [editorRef.current, onUpdate]);

  const onDataUrlChange = (value: string) => setImageDataUrl(value);
  const onImageChange = (img: HTMLImageElement) => setImage(img);

  const { onDateInput } = useDate();
  const { onTimeInput } = useTime();
  const { onProgramInput } = useProgram();
  const { onFontFamilyInput } = useFontFamily(editorRef, onFieldChange);
  const { onArtistInput } = useArtist(editorRef, onFieldChange);
  const { onImageInput } = useImage(editorRef, onDataUrlChange, onImageChange);

  useEffect(() => {
    const editor = editorRef.current;
    if (!editor) return;
    preset
      .then(({ assets, labels }) => {
        editor.addAssets(assets);
        editor.setLabels(labels);
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.log(error);
      });
  }, []);

  return (
    <div className={styles.wrapper}>
      <div className={styles.left}>
        <div className={styles.content}>
          <div className={styles.line}>
            <div className={styles.label}>Date:</div>
            <input
              type="text"
              placeholder="31/12"
              defaultValue={date}
              onInput={({ currentTarget: { value } }) => onDateInput(value)}
              className={styles.input}
            />
          </div>
          <div className={styles.line}>
            <div className={styles.label}>Time:</div>
            <input
              type="text"
              placeholder="19:00"
              defaultValue={time}
              onInput={({ currentTarget: { value } }) => onTimeInput(value)}
              className={styles.input}
            />
          </div>
          <div className={styles.line}>
            <div className={styles.label}>Program:</div>
            <select
              defaultValue={program}
              onInput={({ currentTarget: { value } }) => onProgramInput(value)}
              className={styles.input}
            >
              {Object.entries(programs).map(([key, value]) => (
                <option value={key} key={key}>
                  {value.name}
                </option>
              ))}
            </select>
          </div>
          <div className={styles.line}>
            <div className={styles.label}>Artists:</div>
            <textarea
              placeholder="Name..."
              defaultValue={artist}
              onInput={({ currentTarget: { value } }) => onArtistInput(value)}
              rows={5}
              className={classnames(styles.input, styles.textarea)}
            />
          </div>
        </div>
        <div className={styles.design}>
          <div className={styles.line}>
            <div className={styles.label}>Image:</div>{' '}
            <input type="file" ref={fileInputRef} onChange={onImageInput} />
          </div>

          {image && (
            <div className={styles.line}>
              <div className={styles.label}>Image size:</div> {image.width}x{image.height}
            </div>
          )}
          <div className={styles.line}>
            <div className={styles.label}>Font family:</div>
            <input
              type="text"
              placeholder="Arial"
              defaultValue={fontFamily}
              onInput={({ currentTarget: { value } }) => onFontFamilyInput(value)}
              className={styles.input}
            />
          </div>
        </div>
        {image && (
          <div className={styles.output}>
            <a href={imageDataUrl} download={fileName} className={styles.download}>
              Download image
            </a>
          </div>
        )}
      </div>
      <div className={styles.right} ref={coverContainerRef}>
        <canvas id="canvas" className={styles.cover} ref={canvasRef} />
      </div>
    </div>
  );
};
