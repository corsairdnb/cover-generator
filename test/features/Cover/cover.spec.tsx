import React from 'react';
import { shallow } from 'enzyme';
import { Cover } from '../../../app/features/cover/Cover';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn()
}));

describe('Cover', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<Cover />);
    expect(wrapper).not.toBeFalsy();
    // expect(wrapper).toMatchSnapshot();
  });
});
