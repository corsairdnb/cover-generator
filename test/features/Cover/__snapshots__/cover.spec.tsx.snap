// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cover should match snapshot 1`] = `
<div
  className="wrapper"
>
  <div
    className="left"
  >
    <div
      className="content"
    >
      <p>
        Date:
         
        <input
          onInput={[Function]}
          placeholder="31/12"
          type="text"
        />
      </p>
      <p>
        Time: 
        <input
          placeholder="19:00"
          type="text"
        />
      </p>
      <p>
        Program: 
        <input
          placeholder="Xtra"
          type="text"
        />
      </p>
      <p>
        Artists: 
        <textarea
          placeholder="Name..."
        />
      </p>
    </div>
    <div
      className="design"
    >
      <p>
        Logo:
         
        <select>
          <option>
            Logo 1
          </option>
          <option>
            Logo 2
          </option>
          <option>
            Logo 3
          </option>
        </select>
      </p>
      <p>
        Choose File: 
        <input
          id="file"
          onChange={[Function]}
          type="file"
        />
      </p>
      <p>
        Font family:
         
        <input
          onInput={[Function]}
          placeholder="Arial"
          type="text"
        />
      </p>
    </div>
  </div>
  <div
    className="right"
  >
    <canvas
      className="cover"
      id="canvas"
    />
  </div>
</div>
`;
